File: src/deadlock_build_sync/cache_discovery.py
from __future__ import annotations

from pathlib import Path

from .cache_types import (
    CACHE_RELATIVE_PATH,
    DEADLOCK_APP_ID,
    STEAM_ROOT_RELATIVE_PATHS,
    CacheError,
    CacheLocation,
)


def steam_roots(*, home: Path | None = None) -> tuple[Path, ...]:
    base = home or Path.home()
    roots: list[Path] = []
    identities: set[Path] = set()
    for relative_path in STEAM_ROOT_RELATIVE_PATHS:
        candidate = base / relative_path
        identity = candidate.resolve(strict=False)
        if identity in identities:
            continue
        identities.add(identity)
        roots.append(candidate)
    return tuple(roots)


def steam_root() -> Path:
    roots = steam_roots()
    return next((root for root in roots if root.is_dir()), roots[0])


def _steam_accounts(userdata: Path, account_id: int | None) -> list[Path]:
    if account_id is not None:
        return [userdata / str(account_id)]
    if not userdata.is_dir():
        return []
    return [
        path for path in userdata.iterdir() if path.is_dir() and path.name.isdigit()
    ]


def _explicit_cache_location(
    cache_path: Path,
    account_id: int | None,
) -> CacheLocation:
    resolved = cache_path.expanduser().resolve()
    if not resolved.is_file():
        raise CacheError(f"Deadlock cache does not exist: {resolved}")
    try:
        inferred_account = int(resolved.parents[3].name)
        app_directory = resolved.parents[2]
    except (IndexError, ValueError) as error:
        if account_id is None:
            raise CacheError(
                "--account-id is required with a nonstandard --cache-path"
            ) from error
        inferred_account = account_id
        app_directory = resolved.parent.parent.parent
    if account_id is not None and inferred_account != account_id:
        raise CacheError(
            f"cache belongs to account {inferred_account}, not requested account "
            f"{account_id}"
        )
    return CacheLocation(inferred_account, resolved, app_directory)


def _discover_cache_locations(
    roots: tuple[Path, ...],
    account_id: int | None,
) -> list[CacheLocation]:
    candidates: list[CacheLocation] = []
    discovered_paths: set[Path] = set()
    for steam_directory in roots:
        userdata = steam_directory.expanduser() / "userdata"
        for account in _steam_accounts(userdata, account_id):
            candidate = account / DEADLOCK_APP_ID / CACHE_RELATIVE_PATH
            resolved = candidate.resolve(strict=False)
            if candidate.is_file() and resolved not in discovered_paths:
                discovered_paths.add(resolved)
                candidates.append(
                    CacheLocation(
                        account_id=int(account.name),
                        cache_path=resolved,
                        app_directory=(account / DEADLOCK_APP_ID).resolve(),
                    )
                )
    return candidates


def _select_cache_location(
    candidates: list[CacheLocation],
    account_id: int | None,
) -> CacheLocation:
    if not candidates:
        requested = f" for account {account_id}" if account_id is not None else ""
        raise CacheError(f"no Deadlock Steam Cloud cache found{requested}")
    if len(candidates) > 1:
        accounts = ", ".join(
            f"{candidate.account_id} at {candidate.cache_path}"
            for candidate in candidates
        )
        hint = (
            "--cache-path"
            if len({candidate.account_id for candidate in candidates}) == 1
            else "--account-id or --cache-path"
        )
        raise CacheError(f"multiple Deadlock caches found ({accounts}); pass {hint}")
    return candidates[0]


def discover_cache(
    *,
    account_id: int | None = None,
    cache_path: Path | None = None,
    root: Path | None = None,
) -> CacheLocation:
    if cache_path is not None:
        return _explicit_cache_location(cache_path, account_id)
    roots = (root,) if root is not None else steam_roots()
    return _select_cache_location(
        _discover_cache_locations(roots, account_id), account_id
    )


def deadlock_is_running() -> bool:
    proc = Path("/proc")
    for entry in proc.iterdir():
        if not entry.name.isdigit():
            continue
        try:
            command = (
                (entry / "cmdline")
                .read_bytes()
                .replace(b"\0", b" ")
                .decode("utf-8", errors="ignore")
            )
        except (FileNotFoundError, PermissionError, ProcessLookupError):
            continue
        lowered = command.casefold()
        if "deadlock.exe" in lowered and (
            "steamapps/common/deadlock" in lowered or "s:\\common\\deadlock" in lowered
        ):
            return True
    return False

File: crates/deadlock-steam/src/cache_discovery.rs
use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf};

use deadlock_data::{Error, Result};

use crate::cache_location::{APPLICATION_ID, CACHE_RELATIVE_PATH, CacheLocation, CacheSearch};

const STEAM_ROOTS: [&str; 5] = [
    ".local/share/Steam",
    ".steam/steam",
    ".steam/root",
    ".var/app/com.valvesoftware.Steam/.local/share/Steam",
    "snap/steam/common/.local/share/Steam",
];

/// # Errors
/// Returns an error when an existing Steam directory cannot be resolved.
pub fn steam_roots(home: &Path) -> Result<Vec<PathBuf>> {
    let mut roots = Vec::new();
    let mut seen = BTreeSet::new();
    for relative in STEAM_ROOTS {
        let candidate = home.join(relative);
        if candidate.try_exists()? {
            let canonical = candidate.canonicalize()?;
            if canonical.is_dir() && seen.insert(canonical.clone()) {
                roots.push(canonical);
            }
        }
    }
    Ok(roots)
}

/// # Errors
/// Returns an error when no cache matches, multiple caches match, or the account does not match.
pub fn discover_cache(search: &CacheSearch, home: &Path) -> Result<CacheLocation> {
    if search.account_id == Some(0) {
        return Err(Error::new("Steam account ID must be positive"));
    }
    if let Some(path) = &search.cache_path {
        return explicit_location(path, search.account_id);
    }
    let roots = match &search.steam_root {
        Some(root) => vec![root.clone()],
        None => steam_roots(home)?,
    };
    let mut candidates = Vec::new();
    let mut seen = BTreeSet::new();
    for root in roots {
        for (account_id, account) in accounts(&root.join("userdata"), search.account_id)? {
            let app_directory = account.join(APPLICATION_ID);
            let cache_path = app_directory.join(CACHE_RELATIVE_PATH);
            if cache_path.is_file() {
                let cache_path = cache_path.canonicalize()?;
                if seen.insert(cache_path.clone()) {
                    candidates.push(CacheLocation {
                        account_id,
                        cache_path,
                        app_directory: app_directory.canonicalize()?,
                    });
                }
            }
        }
    }
    select_location(candidates)
}

fn accounts(userdata: &Path, selected: Option<u32>) -> Result<Vec<(u32, PathBuf)>> {
    if let Some(account) = selected {
        return Ok(vec![(account, userdata.join(account.to_string()))]);
    }
    if !userdata.try_exists()? {
        return Ok(Vec::new());
    }
    let mut accounts = Vec::new();
    for entry in fs::read_dir(userdata)? {
        let entry = entry?;
        if let Some(account) = entry
            .file_name()
            .to_str()
            .and_then(|name| name.parse::<u32>().ok())
            && account != 0
            && entry.path().is_dir()
        {
            accounts.push((account, entry.path()));
        }
    }
    accounts.sort_unstable();
    Ok(accounts)
}

fn explicit_location(path: &Path, selected: Option<u32>) -> Result<CacheLocation> {
    let cache_path = path
        .canonicalize()
        .map_err(|error| Error::new(format!("Cannot resolve cache {}: {error}", path.display())))?;
    if !cache_path.is_file() {
        return Err(Error::new("Steam cache path must be a regular file"));
    }
    let app_directory = cache_path
        .ancestors()
        .nth(3)
        .ok_or_else(|| Error::new("Steam cache path has no application directory"))?
        .to_path_buf();
    let inferred = infer_account(&app_directory);
    let account_id = selected
        .or(inferred)
        .filter(|account| *account != 0)
        .ok_or_else(|| Error::new("A nonstandard cache path requires --account-id"))?;
    if inferred.is_some_and(|account| account != account_id) {
        return Err(Error::new(
            "Steam cache account does not match --account-id",
        ));
    }
    Ok(CacheLocation {
        account_id,
        cache_path,
        app_directory,
    })
}

fn infer_account(app_directory: &Path) -> Option<u32> {
    if app_directory.file_name()? != APPLICATION_ID {
        return None;
    }
    app_directory.parent()?.file_name()?.to_str()?.parse().ok()
}

fn select_location(mut candidates: Vec<CacheLocation>) -> Result<CacheLocation> {
    if candidates.len() > 1 {
        let descriptions = candidates
            .iter()
            .map(|location| {
                format!(
                    "{} at {}",
                    location.account_id,
                    location.cache_path.display()
                )
            })
            .collect::<Vec<_>>()
            .join(", ");
        return Err(Error::new(format!(
            "Multiple Steam caches match: {descriptions}. Set --account-id or --cache-path"
        )));
    }
    candidates
        .pop()
        .ok_or_else(|| Error::new("No Deadlock Steam cache matches the request"))
}
